mcodingbot/tasks/stats.py

```diff
@@ -0,0 +1,136 @@
+
+LOGGER = logging.getLogger(__file__)
+
+
+async def loop_update_channels(bot: Bot) -> None:
+    while True:
+        try:
+            await update_channels(bot)
+        except Exception:
+            LOGGER.error("Failed to update channel stats:", exc_info=True)
+
+        await asyncio.sleep(5 * 60)
+
+
+async def update_channels(bot: Bot) -> None:
+    stats = await get_stats(bot)
+
+    sub_channel = bot.cache.get_guild_channel(CONFIG.sub_count_channel)
+    view_channel = bot.cache.get_guild_channel(CONFIG.view_count_channel)
+    member_channel = bot.cache.get_guild_channel(CONFIG.member_count_channel)
+
+    if sub_channel:
+        await sub_channel.edit(name=f"Subs: {display_stats(stats.subs)}")
+    else:
+        LOGGER.warning("No sub count channel to update stats for.")
+
+    if view_channel:
+        await view_channel.edit(name=f"Views: {display_stats(stats.views)}")
+    else:
+        LOGGER.warning("No view count channel to update stats for.")
+
+    guild = bot.cache.get_guild(CONFIG.mcoding_server)
+    if not guild:
+        return LOGGER.warning(
+            "Couldn't find mCoding guild, not updating member count."
+        )
+    if not member_channel:
+        return LOGGER.warning("No member count channel to update stats for.")
+    member_count = guild.member_count
+    if member_count is None:
+        return LOGGER.error("Cached guild didn't contain member count.")
+
+    await member_channel.edit(name=f"Members: {display_stats(member_count)}")
+
+
+@dataclass
+class Stats:
+    subs: float
+    views: float
+
+
+BASE_URL = "https://www.googleapis.com/youtube/v3/channels"
+_last_known_stats = Stats(0, 0)
+
+
+async def get_stats(bot: Bot) -> Stats:
+    link = (
+        f"{BASE_URL}?part=statistics&id={CONFIG.mcoding_yt_id}"
+        f"&key={CONFIG.yt_api_key}"
+    )
+
+    async with bot.session.get(link) as res:
+        res.raise_for_status()
+        response: dict[str, Any] = await res.json()
+
+    if not response:
+        LOGGER.error("Received 2XX but no data.")
+        return _last_known_stats
+
+    items = response.get("items")
+    if not items:
+        LOGGER.error("Response did not contain 'items'.")
+        return _last_known_stats
+
+    channel = items[0]
+    if channel.get("id") != CONFIG.mcoding_yt_id:
+        LOGGER.error("Channel ID was not mCoding.")
+        return _last_known_stats
+
+    statistics = channel.get("statistics")
+    if not statistics:
+        LOGGER.error("mCoding channel did not contain 'statistics'.")
+        return _last_known_stats
+
+    subs = float(statistics.get("subscriberCount", 0))
+    views = float(statistics.get("viewCount", 0))
+
+    if not (subs and views):
+        LOGGER.error(
+            "Statistics did not contain subscriberCount and viewCount."
+        )
+        return _last_known_stats
+
+    _last_known_stats.subs = subs
+    _last_known_stats.views = views
+    return _last_known_stats
+
+
+def display_stats(stat: int | float) -> str:
```

Comment:
So we're using py 3.10+ annotations?

Reply:
We have a `from __future__ import annotations` statement, so 3.9+ annotations work in older python versions.

Reply:
oh i didn't knew that, how much is it backwards compatible?

Reply:
min supported version is py 3.8